data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -0,0 +1,238 @@
+      StructLikeSet deleteSet = Deletes.toEqualitySet(
+          // copy the delete records because they will be held in a set
+          CloseableIterable.transform(CloseableIterable.concat(deleteRecords), Record::copy),
+          deleteSchema.asStruct());
+
+      filteredRecords = Deletes.filter(filteredRecords,
+          record -> projectRow.wrap(asStructLike(record)), deleteSet);
+    }
+
+    return filteredRecords;
+  }
+
+  private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
+    if (posDeletes.isEmpty()) {
+      return records;
+    }
+
+    List<CloseableIterable<Record>> deletes = Lists.transform(posDeletes, this::openPosDeletes);
+
+    // if there are fewer deletes than a reasonable number to keep in memory, use a set
+    if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < setFilterThreshold) {
+      return Deletes.filter(
+          records, this::pos,
+          Deletes.toPositionSet(dataFile.path(), CloseableIterable.concat(deletes)));
+    }
+
+    return Deletes.streamingFilter(records, this::pos, Deletes.deletePositions(dataFile.path(), deletes));
+  }
+
+  private CloseableIterable<Record> openPosDeletes(DeleteFile file) {
+    return openDeletes(file, POS_DELETE_SCHEMA);
+  }
+
+  private CloseableIterable<Record> openDeletes(DeleteFile deleteFile, Schema deleteSchema) {
+    InputFile input = getInputFile(deleteFile.path().toString());
+    switch (deleteFile.format()) {
+      case AVRO:
+        return Avro.read(input)
+            .project(deleteSchema)
+            .reuseContainers()
+            .createReaderFunc(DataReader::create)
+            .build();
+
+      case PARQUET:
+        Parquet.ReadBuilder builder = Parquet.read(input)
+            .project(deleteSchema)
+            .reuseContainers()
+            .createReaderFunc(fileSchema -> GenericParquetReaders.buildReader(deleteSchema, fileSchema));
+
+        if (deleteFile.content() == FileContent.POSITION_DELETES) {
+          builder.filter(Expressions.equal(MetadataColumns.DELETE_FILE_PATH.name(), dataFile.path()));
+        }
+
+        return builder.build();
+
+      case ORC:
+      default:
+        throw new UnsupportedOperationException(String.format(
+            "Cannot read deletes, %s is not a supported format: %s", deleteFile.format().name(), deleteFile.path()));
+    }
+  }
+
+  private static Schema fileProjection(Schema tableSchema, Schema requestedSchema,
+                                       List<DeleteFile> posDeletes, List<DeleteFile> eqDeletes) {
+    if (posDeletes.isEmpty() && eqDeletes.isEmpty()) {
+      return requestedSchema;
+    }
+
+    Set<Integer> requiredIds = Sets.newLinkedHashSet();
+    if (!posDeletes.isEmpty()) {
+      requiredIds.add(MetadataColumns.ROW_POSITION.fieldId());
+    }
+
+    for (DeleteFile eqDelete : eqDeletes) {
+      requiredIds.addAll(eqDelete.equalityFieldIds());
+    }
+
+    Set<Integer> missingIds = Sets.newLinkedHashSet(
```

Comment:
Seems there's no need to create another new LinkedHashSet ? we won't modify this missingIds set, right  ?

Reply:
This is mostly to avoid diffing the sets twice, once in `isEmpty` and once to iterate.

Reply:
Okay, that make sense.